src/Features/Core/Portable/Workspace/BackgroundCompiler.cs

```diff
@@ -40,141 +53,107 @@ public BackgroundCompiler(Workspace workspace)
+        private async ValueTask AddCompilationsForVisibleDocumentsAsync(
+            ImmutableSegmentedList<CancellationToken> cancellationTokens,
+            ArrayBuilder<Compilation> compilations,
+            CancellationToken disposalToken)
         {
-            lock (_buildGate)
-            {
-                _cancellationSource.Cancel();
-                _cancellationSource = new CancellationTokenSource();
-                if (releasePreviousCompilations)
-                {
-                    _mostRecentCompilations = null;
-                }
-            }
+            var workspace = _workspace;
+            if (workspace is null)
+                return;
+
+            // Because we always cancel the previous token prior to queuing new work, there can only be at most one
+            // actual real cancellation token that is not already canceled.
+            var cancellationToken = cancellationTokens.SingleOrDefault(ct => !ct.IsCancellationRequested);
+
+            // if we didn't get an actual non-canceled token back, then this batch was entirely canceled and we have
+            // nothing to do.
+            if (cancellationToken == default)
+                return;
+
+            using var source = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, disposalToken);
+            await AddCompilationsForVisibleDocumentsAsync(
+                workspace.CurrentSolution, compilations, source.Token).ConfigureAwait(false);
         }
 
-        private Task BuildCompilationsAsync(
+        private static async ValueTask AddCompilationsForVisibleDocumentsAsync(
             Solution solution,
-            ProjectId? initialProject,
-            ISet<ProjectId> allOpenProjects)
-        {
-            var cancellationToken = _cancellationSource.Token;
-            return _taskQueue.ScheduleTask(
-                "BackgroundCompiler.BuildCompilationsAsync",
-                () => BuildCompilationsAsync(solution, initialProject, allOpenProjects, cancellationToken),
-                cancellationToken);
-        }
-
-        private Task BuildCompilationsAsync(
-            Solution solution,
-            ProjectId? initialProject,
-            ISet<ProjectId> projectsToBuild,
+            ArrayBuilder<Compilation> compilations,
             CancellationToken cancellationToken)
         {
-            var allProjectIds = new List<ProjectId>();
-            if (initialProject != null)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var trackingService = solution.Workspace.Services.GetRequiredService<IDocumentTrackingService>();
+            var visibleProjectIds = trackingService.GetVisibleDocuments().Select(d => d.ProjectId).ToSet();
+            var activeProjectId = trackingService.TryGetActiveDocument()?.ProjectId;
+
+            // Prioritize the project for the active document first.
+            await GetCompilationAsync(activeProjectId).ConfigureAwait(false);
```

Comment:
only processing project one at a time so we don't saturate everything running parallel.